### ISSUE TITLE
Feature Request: API Management support

### ISSUE THREAD

**Issue by pietro** [terraform-providers/terraform-provider-azurerm#1177]
I couldn't find anything on the documentation about API Gateways, policies, etc. for Azure API Management[0][1].

[0] https://docs.microsoft.com/en-us/azure/api-management/
[1] https://docs.microsoft.com/en-us/rest/api/apimanagement/

**Comment by woosley** [comment]
I am looking for the same thing, is this in any plan?

**Comment by torresdal** [comment]
I've started to look at the other relevant `api-management` resources to implement. I would like your input on where to draw the line between IaC (Infrastructure as Code) and deployment. The `api-management`-service is clearly IaC in my opinion and I didn't hesitate to implement  it. Other relevant resources of `api-management` is not necessarily as clear.

I've listed possible `api-management` Terraform resources extracted from the [Azure API Management REST API](https://docs.microsoft.com/en-us/rest/api/apimanagement/). After this exercise I was left with the following questions:

1. Which of these are clearly IaC and should be resources in Terraform? Or which ones are clearly not IaC and does not belong in Terraform?
2. In which priority? *"Without resource x, y, z, we can't configure API Management in Terraform"* 
2. Are some of these resources better handled elsewhere (other tooling?)
3. Should all be separate resources, and does some belong together/merged?

**Note:** The ones with ~~strikethrough~~ are resources I personally think is irrelevant for Terraform.

* `api-management-api`
* `api-management-api-diagnostic`
* ~~`api-management-api-issue`~~
* `api-management-api-logger`
* `api-management-api-operation`
* `api-management-api-operation-policy`
* `api-management-api-policy`
* `api-management-api-release`
* `api-management-api-schema`
* `api-management-api-version-set`
* `api-management-authorization-server`
* `api-management-backend`
* `api-management-certificate`
* `api-management-delegation`
* `api-management-diagnostic`
* `api-management-email-template`
* `api-management-group`
* `api-management-identity-provider`
* ~~`api-management-issue`~~
* `api-management-logger` 
* ~~`api-management-notification`~~
* `api-management-oidc-provider`
* `api-management-policy`
* `api-management-product`
* `api-management-product-group`
* `api-management-product-policy`
* `api-management-product-subscription`
* `api-management-property`
* `api-management-tag`
* `api-management-user`

**Comment by torresdal** [comment]
Hi all! Quick status. Picking this up again after a rather long summer vacation :-) 

Currently the pull request (#1516) for the api management service is awaiting approval (expect this to go out very soon after great reviews @tombuildsstuff @jeffreyCline), and I've started the work on api-management-api. 

@lfshr You mentioned you might be interested in contributing - that would be great! It will probably be easier to do so when both `api-management` and `api-management-api` is merged into master, but I'm open to suggestions. Let me know what you think?

**Comment by torresdal** [comment]
With regards to feature prioritization I totally agree with your assessment @anoff, and I think this is what we should aim for. Btw if anyone wants to contribute to any of these features, that would be very welcome. I've just started with the basic stuff, but too much work for me to do all these features on my own 😄

**Comment by torresdal** [comment]
While working on the `api-management-api` implementation, I'm curious to what people commonly use/need here. Import from Swagger/Wsdl/... or specify each API operation manually (eg GET /users/{user})? What are common usecases?

Allowing users to explicitly define all operations of an API is by all means doable, but will require some work. A very important factor will be to get the abstraction level right, to make it usable.

Any input you have on this would be great. Thanks.

**Comment by alastairtree** [comment]
Import from swagger would be lovely - so that I could just check in a swagger file with some .tf files and apply those rather than a giant pile of ARM template JSON

**Comment by torresdal** [comment]
@alastairtree Import is what I've implemented so far and will probably be what we support in the first release. 

The more I think about it, the more I think being able to add individual operations is not a good idea. If someone have good arguments for the opposite, please let me know.

**Comment by alastairtree** [comment]
Yeah until terraform import can actually generates the HCL in .tf files for you you are probably right. But the ability to define your api mangagement config in HCL with references to swagger files for the api definition would be an nice feature I think.

Any idea on an ETA on API management support in the azurerm terraform provider?

**Comment by svscorp** [comment]
@alastairtree as I see it was merged 7 days ago: MR #1516

Hopefully updates will follow :)

**Comment by acarroll** [comment]
I've implemented the CRUD for product on the side (separate provider), have decided to merge it back into the main provider and am looking to complete that and raise a pull request hopefully this week, if I can get enough time.

For the project on which I'm currently working, the things that're candidates for TF provisioning are the APIMS, products, product policies, name value pairs and API version sets.  These are the things we don't consider to be very volatile.  We're often redeploying Swagger specs and versions during development so from this perspective I haven't considered API as a TF candidate.

**Comment by torresdal** [comment]
Just letting you know that I wont be contributing any more to this repository and the API Management resources, until I see major improvements in lead times. After months of coding resources for Azure API Management, only the first resource is merged (api_management). My second (api_management_api) has been in review for 2 months and nothing is happening and nobody is answering my questions - so I’ve now closed it (https://github.com/terraform-providers/terraform-provider-azurerm/pull/2073). Hope other contributors gets a better experience contributing here, because I’m done wasting my time.

**Comment by mcwienczek** [comment]
Can someone from Microsoft address why this PR couldn't be merged? And most importantly when those features will land in terraform then?

**Comment by torresdal** [comment]
Also interested In that. 

@mcwienczek I guess that question should go to Hashicorp and not Microsoft, since this is a Hashicorp open source project.

**Comment by michaelmeelis** [comment]
I would like to see api-management-settings-virtual-network. I'm using the apim in combination with AKS and I need to put the APIM into an existing vnet->subnet. Is this a different issue or does this belong to this feature request.

**Comment by AndyMoore** [comment]
so in the absence of this, we're looking at either embedding arm into terraform configuration, or just moving to aws?

**Comment by tombuildsstuff** [comment]
hey @torresdal 

So firstly I apologies that it's taken so long to reply to this.

Spending some time thinking about this issue, based on the comments above I think this ultimately comes down to 3 questions:

1. Why did this PR sit for so long?
2. When will these features be landing?
3. What are we doing to improve this?

So:

> Why did this PR sit for so long?

This is mostly unintentional, from our side we were busy with other things (personally that's some holiday, HashiConf and then [the Requiring Imports changes](https://github.com/terraform-providers/terraform-provider-azurerm/issues/2807)) - and due to taking a quick skim over this PR I believed this was still waiting on changes to be made at your end. In either case unfortunately this is my fault for missing that you'd pushed changes here - so apologies about that.

Whilst I appreciate that probably isn't helpful at this point - we're aware there are pain points in the current flow (and I'll cover more about and our plan to improve that below).

> When will these features be landing?

So I have good news on this front, we're going to be working on adding support for the following API Management Resources in the next few releases:

* Products: `azurerm_api_management_product` (PR: #2953)
* Users: `azurerm_api_management_user` (PR: #2954)
* Groups: `azurerm_api_management_group` (PR: #2809)
* Group Users: `azurerm_api_management_group_user` (working title)
* Product Groups: `azurerm_api_management_product_group` (working title)
* APIs: `azurerm_api_management_api` (working title)
* Product APIs: `azurerm_api_management_product_api` (working title)
* Subscriptions: `azurerm_api_management_subscription` (working title)
* Property: `azurerm_api_management_property` (working title)
* API Operations: `azurerm_api_management_api_operation` (working title)
* Authorization Servers: `azurerm_api_management_authorization_server` (working title)
* Certificates: `azurerm_api_management_certificate` (working title)
* Loggers: `azurerm_api_management_logger` (working title)
* Open ID Connect Providers: `azurerm_api_management_openid_connect_provider` (working title)
* Policy: `azurerm_api_management_policy` (working title)

Whilst I don't have an exact timeline for each Resource, after spending some time trying to work out the dependencies between these resources - this is the (more or less) ordered list we're planning on supporting over the next few releases.

In addition we also plan to enhance the `azurerm_api_management` resource to add support for some of the fields which aren't currently supported.

> What are we doing to improve this?

Whilst I appreciate this experience has been frustrating - we're aware there's issues that require fixing here and we're working to improve this going forward.

One of the benefits of investing heavily in the Azure Provider over the last couple of years has been the marked increase in adoption, both in terms of folks using the provider but also in terms of issues and PR's.

Whilst on the one hand we've done a fairly good job in terms of adding support for new resources - on the other hand we've not done so well in terms of bug fixes or community PR's - and we're intending on fixing that. Unfortunately for historical reasons our roadmaps pretty fixed until the end of April - but from that point we're intending on switching the ratios such that we can spend more time on Community PR's, Bug Fixes and Issues (as well as new resource development) - rather than being solely focused on new resource development.

Taking a look at where we are today, we're trying to review PR's and get them merged within a day or two of them coming in, which tends to work for the smaller PR's. On the flip side some of the larger PR's can take longer and sometimes fall through the cracks - in particular where they require more thought (such as #1805) or have an upstream dependency blocking them (such as #2582).

From our side @katbyte has spent some time adding linters to try and ensure the low-hanging comments are covered - and ensuring the codebase is a little more consistent insofar as when users look to existing resources, they're not using a pattern we're deprecating. In addition, we've been trying to ensure the PR reviews that we do are comprehensive - which whilst this may introduce a larger number of comments does reduce the amount of back and forth so that we can get things merged faster. Finally when reviewing PR's that are otherwise good to merge (for example we've not heard back from a user) we've been pushing the odd commit to the branch so that we can get these merged.

Over the past few months we've been working to try and make what's being worked on more visible through the project's GitHub Milestones, both in terms of the work we're planning on doing and in terms of PR's we plan to review.

This includes the use of the 'blocked' milestone where appropriate (e.g. where a PR is blocked on a broken API such as #2582) - where we've recently switched tact to temporarily closing PR's which are blocked on an upstream issue but leaving them assigned to the `blocked` milestone, which allows us to focus on other tasks whilst the upstream issues are resolved, rather than having them left open indefinitely.

Thinking ahead slightly further, we're also looking to onboard additional maintainers for the Azure Provider - but I don't have a timeframe for this unfortunately.

In the case of this particular Pull Request - apologies that this has slipped through the cracks, this was an error on my part. If you're still interested in contributing support for API Management API's we'd love to take another look and get this over the line - in either case I'd like to thank you for this contribution even though I appreciate it's been a frustrating experience.

Thanks!

**Comment by michaelmeelis** [comment]
Hi @tombuildsstuff,

Great to hear that this part is getting some well deserved love. I'm still curious if the vnet part will be picked up as well. I've currently deployed the APIM with an ARM template to circumvent this problem but would love to see this back into this resource. 

Kind regards,
Michael

**Comment by tombuildsstuff** [comment]
@michaelmeelis support for API Management Virtual Networks is being implemented in https://github.com/terraform-providers/terraform-provider-azurerm/pull/2582 - however it's blocked on an upstream issue, which is why that PR is temporarily closed

**Comment by michaelmeelis** [comment]
@tombuildsstuff thanks for the quick response! I will look at that issue.

**Comment by NathanielRose** [comment]
Hey @tombuildsstuff,

While testing the Azure API Management Service Provider I noticed a few things. 

- The terraform output from a provisioned API-mgmt service returns an scm_url with no credentials. This requires us to run a separate script to generate the tokens for us to push to the repo.
- Once keys are generated you then have to deploy your commit to the API-mgmt service through the rest api adding another step in an automated script.
- Is there a way to point the scm_url to a pre-existing git repo? (More of an API-mgmt service request)

**Comment by tombuildsstuff** [comment]
@NathanielRkse would you mind opening a separate issue for those? Thanks!

**Comment by torresdal** [comment]
For the record, thanks for coming back with an explanation @tombuildsstuff!  Looks like the activity on APIM has really picked up and things are moving. Good to see!
